### PR TITLE
Porting to llvm-18

### PR DESCRIPTION
**Pull request by sjalander**
Clang format seems to have caused some files to be formatted differently, which makes the diff larger than necessary.

### CODE REVIEW

File: tests/jlm/llvm/frontend/llvm/LoadTests.cpp [hunk 1]
```diff
@@ -25,7 +25,8 @@ LoadConversion()
   std::unique_ptr<Module> llvmModule(new Module("module", context));
 
   auto int64Type = Type::getInt64Ty(context);
-  auto pointerType = Type::getInt64PtrTy(context);
```
Comment (sjalander) [inline tests/jlm/llvm/frontend/llvm/LoadTests.cpp:28]: The Int64PtrTy seems to be deprecated.

File: tests/jlm/llvm/frontend/llvm/LoadTests.cpp [hunk 2]
```diff
@@ -25,7 +25,8 @@ LoadConversion()
...
+  // ADDRESS_SPACE_GENERIC == 0
+  auto pointerType = llvm::PointerType::get(Type::getInt64Ty(context), 0);
```
Comment (sjalander) [inline tests/jlm/llvm/frontend/llvm/LoadTests.cpp:29]: This was what I managed to find as a replacement that works for the test.
The strange part is that the AddressSpace enum seems to be related to Nvidia PTX, which is the reason that I pass in '0'.

File: tests/jlm/llvm/frontend/llvm/MemCpyTests.cpp [hunk 3]
```diff
@@ -25,7 +25,8 @@ MemCpyConversion()
   std::unique_ptr<Module> llvmModule(new Module("module", context));
 
   auto int64Type = Type::getInt64Ty(context);
-  auto pointerType = Type::getInt64PtrTy(context);
```
Comment (sjalander) [inline tests/jlm/llvm/frontend/llvm/MemCpyTests.cpp:28]: Same as above

**Comment by phate** [comment]
@sjalander Could you put the change in the CI where you introduce the `llvm-version` input to the installation script into its own PR? It has nothing really to do with porting to LLVM 18, but would be nice to have.

File: tests/jlm/llvm/frontend/llvm/LoadTests.cpp (hunk 2 above)
Comment (phate) [reply tests/jlm/llvm/frontend/llvm/LoadTests.cpp:29]: I think the `PointerType::getUnqual()` is the right one to use here: https://llvm.org/doxygen/classllvm_1_1PointerType.html

File: tests/jlm/llvm/frontend/llvm/MemCpyTests.cpp (hunk 3 above)
Comment (phate) [reply tests/jlm/llvm/frontend/llvm/MemCpyTests.cpp:28]: same as above

**Review by phate — COMMENTED** [review]
@haved @caleridas Any objections for moving over to LLVM 18?

**Comment by phate** [comment]
@haved @caleridas Any objections to moving to LLVM 18?

**Comment by haved** [comment]
> @haved @caleridas Any objections to moving to LLVM 18?

Nope, I'm all for it.

**Comment by phate** [comment]
@sjalander Could finish up this PR such that we can merge it, please?

**Comment by sjalander** [comment]
@phate
What is it that you want fixed?

File: .github/actions/InstallPackages/action.yml [hunk 4]
```diff
@@ -85,8 +85,8 @@ runs:
       run: |
         sudo apt-get install libmlir-${{inputs.llvm-version}}-dev mlir-${{inputs.llvm-version}}-tools
         if ! [ -f /usr/lib/x86_64-linux-gnu/libMLIR.so ]; then
-          sudo ln -s /usr/lib/llvm-${{inputs.llvm-version}}/lib/libMLIR.so.${{inputs.llvm-version}} /usr/lib/x86_64-linux-gnu/
-          sudo ln -s /usr/lib/llvm-${{inputs.llvm-version}}/lib/libMLIR.so.${{inputs.llvm-version}} /usr/lib/x86_64-linux-gnu/libMLIR.so
+          sudo ln -s /usr/lib/llvm-${{inputs.llvm-version}}/lib/libMLIR.so.${{inputs.llvm-version}}.1 /usr/lib/x86_64-linux-gnu/
+          sudo ln -s /usr/lib/llvm-${{inputs.llvm-version}}/lib/libMLIR.so.${{inputs.llvm-version}}.1 /usr/lib/x86_64-linux-gnu/libMLIR.so
```
[the remark this answers is not in the document]
Comment (phate) [reply .github/actions/InstallPackages/action.yml:89]: @sjalander You tried to remove them, but the HLS test suite CI failed, right? What was the problem there?

File: scripts/run-hls-test.sh [hunk 5]
```diff
@@ -5,6 +5,7 @@ set -eu
 GIT_REPOSITORY=https://github.com/phate/hls-test-suite.git
 GIT_COMMIT=effbe0bff96b396fb41e7c95bb74c7c772567136
 
+
```
Comment (phate) [inline scripts/run-hls-test.sh:8]: Noise

File: tests/jlm/llvm/frontend/llvm/StoreTests.cpp [hunk 6]
```diff
@@ -25,7 +25,7 @@ StoreConversion()
   std::unique_ptr<Module> llvmModule(new Module("module", context));
 
   auto int64Type = Type::getInt64Ty(context);
-  auto pointerType = Type::getInt64PtrTy(context);
+  auto pointerType = llvm::PointerType::getUnqual(Type::getInt64Ty(context));
```
Comment (phate) [inline tests/jlm/llvm/frontend/llvm/StoreTests.cpp:28]: I think the `PointerType::getUnqual()` without the type argument(!) is the right one to use here: https://llvm.org/doxygen/classllvm_1_1PointerType.html

**Review by phate — CHANGES_REQUESTED** [review]
@sjalander The comments that are still open (or were closed without actually fixing the issue).

**Comment by haved** [comment]
@phate There might be something weird going on with GitHub and responses to comments, because I am fairly certain I saw on @sjalander's screen that he had written several responses to comments that I am not seeing on my own devices. For example regarding the `.1` suffix for `libMLIR.so`. I guess he could have deleted those responses, but I don't see why he would.

**Comment by sjalander** [comment]
@phate
Anything else to fix?

**Comment by haved** [comment]
@phate GitHub is behaving quite strangely. I can see all of @sjalander's comments now, but they are all independent of the comments they reply to. They also show an older version of the changes made in this PR, without having the "outdated" tag that I usually expect when a comment is made on an outdated diff

File: tests/jlm/llvm/frontend/llvm/MemCpyTests.cpp (hunk 3 above)
Comment (phate) [reply tests/jlm/llvm/frontend/llvm/MemCpyTests.cpp:28]: Yes, but it is still the one with the type argument, and not the one without.

File: tests/jlm/llvm/frontend/llvm/LoadTests.cpp (hunk 1 above)
Comment (phate) [reply tests/jlm/llvm/frontend/llvm/LoadTests.cpp:28]: The same here. It is still the one with the type argument.

File: tests/jlm/llvm/frontend/llvm/LoadTests.cpp [hunk 7]
```diff
@@ -25,7 +25,8 @@ LoadConversion()
   std::unique_ptr<Module> llvmModule(new Module("module", context));
 
   auto int64Type = Type::getInt64Ty(context);
-  auto pointerType = Type::getInt64PtrTy(context);
+  // ADDRESS_SPACE_GENERIC == 0
```
Comment (phate) [inline tests/jlm/llvm/frontend/llvm/LoadTests.cpp]: @sjalander Where does this comment come from all in a sudden?

**Review by phate — CHANGES_REQUESTED** [review]

File: tests/jlm/llvm/frontend/llvm/LoadTests.cpp (hunk 7 above)
Comment (sjalander) [reply tests/jlm/llvm/frontend/llvm/LoadTests.cpp]: It's outdated

File: tests/jlm/llvm/frontend/llvm/MemCpyTests.cpp [hunk 8]
```diff
@@ -25,7 +25,7 @@ MemCpyConversion()
...
+  auto pointerType = llvm::PointerType::getUnqual(Type::getInt64Ty(context));
```
[the remark this answers is not in the document]
Comment (sjalander) [reply tests/jlm/llvm/frontend/llvm/MemCpyTests.cpp]: It now uses getUnqual(context).

File: tests/jlm/llvm/frontend/llvm/LoadTests.cpp [hunk 9]
```diff
@@ -25,7 +25,7 @@ LoadConversion()
   std::unique_ptr<Module> llvmModule(new Module("module", context));
 
   auto int64Type = Type::getInt64Ty(context);
-  auto pointerType = Type::getInt64PtrTy(context);
+  auto pointerType = llvm::PointerType::getUnqual(Type::getInt64Ty(context));
```
[the remark this answers is not in the document]
Comment (sjalander) [reply tests/jlm/llvm/frontend/llvm/LoadTests.cpp]: It now uses getUnqual(context).

**Review by phate — APPROVED** [review]

**Pull request merged**